File: sdks/local-rp/php/tests/conformance/KeysTest.php
<?php

declare(strict_types=1);

require_once __DIR__ . '/../bootstrap.php';

use LinkKeys\LocalRp\Crypto;

$vectors = loadJson(__DIR__ . '/../../../conformance/keys.json');

TestKit::test('keys.local_rp.signing.seed_derives_public_key_and_fingerprint', function () use ($vectors) {
    $k = $vectors['local_rp']['signing'];
    $seed = hexToBytes($k['seed_hex']);
    TestKit::assertEquals(hexToBytes($k['private_key_hex']), $seed, 'seed == private_key for Ed25519');
    $pub = Crypto::ed25519PublicKeyFromSeed($seed);
    TestKit::assertEquals(hexToBytes($k['public_key_hex']), $pub);
    TestKit::assertEquals($k['fingerprint_hex'], Crypto::fingerprint($pub));
});

TestKit::test('keys.local_rp.encryption.private_derives_public', function () use ($vectors) {
    $k = $vectors['local_rp']['encryption'];
    $priv = hexToBytes($k['private_key_hex']);
    $pub = Crypto::x25519PublicFromPrivate($priv);
    TestKit::assertEquals(hexToBytes($k['public_key_hex']), $pub);
});

TestKit::test('keys.domain_signing_key.seed_derives_public_key_and_fingerprint', function () use ($vectors) {
    $k = $vectors['domain_signing_key'];
    $seed = hexToBytes($k['seed_hex']);
    $pub = Crypto::ed25519PublicKeyFromSeed($seed);
    TestKit::assertEquals(hexToBytes($k['public_key_hex']), $pub);
    TestKit::assertEquals($k['fingerprint_hex'], Crypto::fingerprint($pub));
});

TestKit::test('keys.domain_encryption_recipient.private_derives_public', function () use ($vectors) {
    $k = $vectors['domain_encryption_recipient'];
    $priv = hexToBytes($k['private_key_hex']);
    $pub = Crypto::x25519PublicFromPrivate($priv);
    TestKit::assertEquals(hexToBytes($k['public_key_hex']), $pub);
});

exit(TestKit::summary('KeysTest') === 0 ? 0 : 1);

File: sdks/local-rp/dart/test/conformance/keys_test.dart
import 'dart:convert';

import 'package:linkkeys_local_rp/linkkeys_local_rp.dart';
import 'package:linkkeys_local_rp/src/crypto/crypto.dart';
import 'package:test/test.dart';

import '../testutil/fixtures.dart';

void main() {
  group('keys.json conformance', () {
    test('fingerprints round trip through SDK fingerprint helpers', () async {
      final d = loadJson('keys.json') as Map<String, dynamic>;
      final nodes = [
        (d['local_rp'] as Map<String, dynamic>)['signing']
            as Map<String, dynamic>,
        d['domain_signing_key'] as Map<String, dynamic>,
      ];

      for (final node in nodes) {
        final seed = hex(node['seed_hex'] as String);
        final expectedPublic = hex(node['public_key_hex'] as String);
        final expectedFp = node['fingerprint_hex'] as String;

        final message = utf8.encode('conformance-check');
        final sig = await Crypto.signEd25519(message, seed);
        expect(
            await Crypto.verifyEd25519(message, sig, expectedPublic), isTrue);

        final computed = await Crypto.fingerprint(expectedPublic);
        expect(computed, equals(expectedFp));

        final s = fingerprintToString(computed);
        expect(fingerprintFromString(s), equals(expectedFp));

        expect(seed.length, equals(32));
      }

      expect(() => fingerprintFromString('deadbeef'),
          throwsA(isA<SdkException>()));
    });

    test('X25519 public keys derive from private keys', () async {
      final d = loadJson('keys.json') as Map<String, dynamic>;
      final nodes = [
        (d['local_rp'] as Map<String, dynamic>)['encryption']
            as Map<String, dynamic>,
        d['domain_encryption_recipient'] as Map<String, dynamic>,
      ];
      for (final node in nodes) {
        final priv = hex(node['private_key_hex'] as String);
        final expectedPublic = hex(node['public_key_hex'] as String);
        final got = await Crypto.derivePublicFromX25519Private(priv);
        expect(got, equals(expectedPublic));
      }
    });
  });
}
